clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/OperationsService.java

```diff
@@ -0,0 +1,252 @@
+                          } else {
+                            // TODO(rz): Assert description is T
+                            validator.validate(descriptions, description, errors);
+                          }
+
+                          allowedAccountValidators.forEach(
+                              it -> {
+                                it.validate(username, allowedAccounts, description, errors);
+                              });
+
+                          // TODO(rz): Assert `description` is T
+                          descriptionAuthorizer.authorize(description, errors);
+
+                          AtomicOperation atomicOperation =
+                              converter.convertOperation(descriptionInput);
+                          if (atomicOperation == null) {
+                            throw new AtomicOperationNotFoundException(descriptionName);
+                          }
+
+                          if (errors.hasErrors()) {
+                            registry
+                                .counter(
+                                    validationErrorsCounterId.withTag(
+                                        "operation", atomicOperation.getClass().getSimpleName()))
+                                .increment();
+                          }
+
+                          return new AtomicOperationBindingResult(atomicOperation, errors);
+                        }))
+        .collect(Collectors.toList());
+  }
+
+  private ProviderVersion getOperationVersion(OperationInput operation) {
+    final String accountName = operation.computeAccountName();
+    if (accountName == null) {
+      log.warn("Unable to get account name from operation: {}", operation);
+    } else {
+      try {
+        AccountCredentials credentials = accountCredentialsRepository.getOne(accountName);
+        return credentials.getProviderVersion();
+      } catch (Exception e) {
+        log.warn("Unable to determine provider version for account {}", accountName, e);
+      }
+    }
+    return ProviderVersion.v1;
+  }
+
+  private static Map processDescriptionInput(
+      Collection<AtomicOperationDescriptionPreProcessor> descriptionPreProcessors,
+      AtomicOperationConverter converter,
+      Map descriptionInput) {
+
+    Method convertDescriptionMethod;
+    try {
+      convertDescriptionMethod = converter.getClass().getMethod("convertDescription", Map.class);
+    } catch (NoSuchMethodException e) {
+      throw new SystemException("Could not find convertDescription method on converter", e);
+    }
+
+    Class<?> convertDescriptionReturnType =
+        ResolvableType.forMethodReturnType(convertDescriptionMethod).getRawClass();
+
+    for (AtomicOperationDescriptionPreProcessor preProcessor : descriptionPreProcessors) {
+      if (preProcessor.supports(convertDescriptionReturnType)) {
+        descriptionInput = preProcessor.process(descriptionInput);
+      }
+    }
+
+    return descriptionInput;
```

Comment:
updated at every step of the loop so we just return the last one?

Reply:
No, the preprocessors mutate the map in a chain. It's the same map the whole way through, for better or worse.

Reply:
ooh I see it's on both sides of the expression